src/toil/spark/spark_cluster.py

```diff
@@ -0,0 +1,328 @@
+            
+
+    def start(self, fileStore):
+        """
+        Start spark and hdfs master containers
+
+        fileStore: Unused
+        """
+        
+        self.IP = subprocess.check_output(["hostname", "-f",])[:-1]
+
+        _log.info("Started Spark master container.")
+        self.sparkContainerID = _docker_call(no_rm = True,
+                                             work_dir = os.getcwd(),
+                                             tool = "quay.io/ucsc_cgl/apache-spark-master:1.5.2",
+                                             docker_parameters = ["--net=host",
+                                                                  "-d",
+                                                                  "-v", "/mnt/ephemeral/:/ephemeral/:rw",
+                                                                  "-e", "SPARK_MASTER_IP="+self.IP,
+                                                                  "-e", "SPARK_LOCAL_DIRS=/ephemeral/spark/local",
+                                                                  "-e", "SPARK_WORKER_DIR=/ephemeral/spark/work"],
+                                             tool_parameters = [],
+                                             sudo = self.sudo,
+                                             check_output = True)[:-1]
+        _log.info("Started HDFS Datanode.")
+        self.hdfsContainerID = _docker_call(no_rm = True,
+                                            work_dir = os.getcwd(),
+                                            tool = "quay.io/ucsc_cgl/apache-hadoop-master:2.6.2",
+                                            docker_parameters = ["--net=host",
+                                                                 "-d"],
+                                            tool_parameters = [self.IP],
+                                            sudo = self.sudo,
+                                            check_output = True)[:-1]
+
+        # attach worker service jobs
+        self._add_worker_services()
+
+        return self.IP
+
+
+    def stop(self, fileStore):
+        """
+        Stop and remove spark and hdfs master containers
+
+        fileStore: Unused
+        """
+        
+        sudo = []
+        if self.sudo:
+            sudo = ["sudo"]
+            
+        subprocess.call(sudo + ["docker", "exec", self.sparkContainerID, "rm", "-r", "/ephemeral/spark"])
+        subprocess.call(sudo + ["docker", "stop", self.sparkContainerID])
+        subprocess.call(sudo + ["docker", "rm", self.sparkContainerID])
+        _log.info("Stopped Spark master.")
+
+        subprocess.call(sudo + ["docker", "stop", self.hdfsContainerID])
+        subprocess.call(sudo + ["docker", "rm", self.hdfsContainerID])
+        _log.info("Stopped HDFS namenode.")
+
+        return
+
+
+    def check(self):
+        """
+        Checks to see if Spark master and HDFS namenode are still running.
+        """
+
```

Comment:
If the service stopping is unexpected and caused by an error the check method should throw an exception rather than returning false. If the service has just.finished.in.an expected.manner returning false is the right thing to.do.